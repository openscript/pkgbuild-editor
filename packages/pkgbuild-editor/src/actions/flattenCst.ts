import { CstElement, CstNode, IToken } from 'chevrotain';

export function flattenCst(cstNode: CstNode): IToken[] {
  const nodes: IToken[] = [];

  function collectNodes(node: CstElement): void {
    if ('children' in node) {
      for (const key in node.children) {
        for (const child of node.children[key]) {
          collectNodes(child);
        }
      }
    } else {
      nodes.push(node);
    }
  }

  collectNodes(cstNode);
  return nodes;
}
