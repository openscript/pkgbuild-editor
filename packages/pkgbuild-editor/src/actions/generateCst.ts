import { ILexingError, IRecognitionException } from 'chevrotain';
import { PkgbuildLexer } from '../lexer';
import { PkgbuildParser } from '../parser';
import { PkgbuildCstNode } from '../parser.gen';

type CstReturn = {
  cst: PkgbuildCstNode | undefined;
  lexerErrors: ILexingError[];
  parserErrors: IRecognitionException[];
};

export const generateCst = (input: string): CstReturn => {
  const lexer = PkgbuildLexer.tokenize(input);
  const parser = new PkgbuildParser(lexer.tokens);
  const cst = parser.pkgbuild();
  return {
    cst,
    lexerErrors: lexer.errors,
    parserErrors: parser.errors,
  };
};
