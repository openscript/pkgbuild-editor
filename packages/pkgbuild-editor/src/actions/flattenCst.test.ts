import { generateCst } from './generateCst';
import { flattenCst } from './flattenCst';

describe('flattenCst.ts', () => {
  it('should flatten Cst', () => {
    const { cst } = generateCst(`# Test comment
varible = "value"`);
    const flattened = flattenCst(cst);
    expect(flattened).toMatchSnapshot();
  });
});
