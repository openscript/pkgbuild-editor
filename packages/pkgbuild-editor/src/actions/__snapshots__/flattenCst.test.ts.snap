// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flattenCst.ts > should flatten Cst 1`] = `
[
  {
    "endColumn": 14,
    "endLine": 1,
    "endOffset": 13,
    "image": "# Test comment",
    "startColumn": 1,
    "startLine": 1,
    "startOffset": 0,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /#\\.\\*/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Comment",
      "tokenTypeIdx": 4,
    },
    "tokenTypeIdx": 4,
  },
  {
    "endColumn": 15,
    "endLine": 1,
    "endOffset": 14,
    "image": "
",
    "startColumn": 15,
    "startLine": 1,
    "startOffset": 14,
    "tokenType": {
      "CATEGORIES": [
        {
          "CATEGORIES": [],
          "PATTERN": /NOT_APPLICABLE/,
          "categoryMatches": [
            5,
            6,
          ],
          "categoryMatchesMap": {
            "5": true,
            "6": true,
          },
          "isParent": true,
          "name": "Formatting",
          "tokenTypeIdx": 3,
        },
      ],
      "LINE_BREAKS": true,
      "PATTERN": /\\\\n/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Newline",
      "tokenTypeIdx": 5,
    },
    "tokenTypeIdx": 5,
  },
  {
    "endColumn": 7,
    "endLine": 2,
    "endOffset": 21,
    "image": "varible",
    "startColumn": 1,
    "startLine": 2,
    "startOffset": 15,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /\\[a-zA-Z_\\]\\[a-zA-Z0-9_\\]\\*/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Variable",
      "tokenTypeIdx": 7,
    },
    "tokenTypeIdx": 7,
  },
  {
    "endColumn": 8,
    "endLine": 2,
    "endOffset": 22,
    "image": " ",
    "startColumn": 8,
    "startLine": 2,
    "startOffset": 22,
    "tokenType": {
      "CATEGORIES": [
        {
          "CATEGORIES": [],
          "PATTERN": /NOT_APPLICABLE/,
          "categoryMatches": [
            5,
            6,
          ],
          "categoryMatchesMap": {
            "5": true,
            "6": true,
          },
          "isParent": true,
          "name": "Formatting",
          "tokenTypeIdx": 3,
        },
      ],
      "PATTERN": /\\\\s\\+/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Whitespace",
      "tokenTypeIdx": 6,
    },
    "tokenTypeIdx": 6,
  },
  {
    "endColumn": 10,
    "endLine": 2,
    "endOffset": 24,
    "image": " ",
    "startColumn": 10,
    "startLine": 2,
    "startOffset": 24,
    "tokenType": {
      "CATEGORIES": [
        {
          "CATEGORIES": [],
          "PATTERN": /NOT_APPLICABLE/,
          "categoryMatches": [
            5,
            6,
          ],
          "categoryMatchesMap": {
            "5": true,
            "6": true,
          },
          "isParent": true,
          "name": "Formatting",
          "tokenTypeIdx": 3,
        },
      ],
      "PATTERN": /\\\\s\\+/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Whitespace",
      "tokenTypeIdx": 6,
    },
    "tokenTypeIdx": 6,
  },
  {
    "endColumn": 9,
    "endLine": 2,
    "endOffset": 23,
    "image": "=",
    "startColumn": 9,
    "startLine": 2,
    "startOffset": 23,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /=/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "Equals",
      "tokenTypeIdx": 9,
    },
    "tokenTypeIdx": 9,
  },
  {
    "endColumn": 17,
    "endLine": 2,
    "endOffset": 31,
    "image": ""value"",
    "startColumn": 11,
    "startLine": 2,
    "startOffset": 25,
    "tokenType": {
      "CATEGORIES": [],
      "PATTERN": /"\\\\w\\+"/,
      "categoryMatches": [],
      "categoryMatchesMap": {},
      "isParent": false,
      "name": "StringLiteral",
      "tokenTypeIdx": 14,
    },
    "tokenTypeIdx": 14,
  },
]
`;
